### DISCUSSION TITLE
Filtering regarding motion artifacts

### DISCUSSION THREAD

**Discussion by Astronomia-maker** [ifm/ifm3d#429] [Q&A]
Hello. 

I have O3R225 which mounted on AGV for our application. So i am currently facing an issue with proper filter setting of pointcloud.

I have two test scenario for the same object. 

Test setup:-

2x Camera O3R225 on AGV. One is 63cm above ground and another is 1.78m above the ground. Both cameras view direction is same.

Test scenario 1:- AGV is not in drive mode stand still. So after cropping in interested region i am getting about 150 point (reducing by voxel size:- 0.02) for my pallet truck (object). Object is about 3m away from the camera. But this pallet truck has shiny surface so i am not getting any pointcloud for top shiny surface.  How can i get pointcloud for this shiny surface also.

Test scenario 2:- AGV drives about 1.5 m/s and camera runs at 20 fps. But this i am getting only 30-40 points when object is about 1.4m near the AGV and when object is 3m aways i have about 10-20 points. And same issue regarding shiny surface.

So i am not getting pointcloud from cameras are in motions. Mostly i have pointcloud when camera is about 1.5m near the object. 

Software version:- 

- IFM firmware:- 1.0.14
- ifm3dpy ==1.3.3

below is my camera configuration.

`
{
    "applications": {
        "classes": {
            "ods": {
                "instanceSchema": {},
                "license": "UNLICENSED",
                "minimumFirmwareVersion": {
                    "major": 0,
                    "minor": 16,
                    "patch": 3
                },
                "version": {
                    "major": 0,
                    "minor": 0,
                    "patch": 0
                }
            }
        }
    },
    "device": {
        "clock": {
            "currentTime": 1581093180309812128,       
            "sntp": {
                "active": true,
                "activeServerAddress": "",
                "activeServerName": "time3.google.com",
                "availableServers": [],
                "systemClockSynchronized": false      
            }
        },
        "diagnostic": {
            "confInitStages": [
                "device",
                "ports",
                "applications"
            ],
            "temperatures": [
                {
                    "entity": "PLL-therm",
                    "value": 50.5
                },
                {
                    "entity": "MCPU-therm",
                    "value": 50.5
                },
                {
                    "entity": "PMIC-Die",
                    "value": 100.0
                },
                {
                    "entity": "Tboard_tegra",
                    "value": 44.0
                },
                {
                    "entity": "GPU-therm",
                    "value": 45.0
                },
                {
                    "entity": "BCPU-therm",
                    "value": 50.5
                },
                {
                    "entity": "thermal-fan-est",      
                    "value": 48.20000076293945        
                },
                {
                    "entity": "Tdiode_tegra",
                    "value": 44.75
                },
                {
                    "entity": "port2",
                    "overtemperature": false,
                    "temperatureLimit": 77,
                    "valid": true,
                    "value": 54
                },
                {
                    "entity": "port3",
                    "overtemperature": false,
                    "temperatureLimit": 77,
                    "valid": true,
                    "value": 64
                }
            ],
            "upTime": 2535000000000
        },
        "info": {
            "description": "",
            "device": "0302",
            "deviceTreeBinaryBlob": "tegra186-quill-p3310-1000-c03-00-base.dtb",
            "features": {},
            "guiSettings": "",
            "name": "",
            "partNumber": "OVP800",
            "productionState": "AA",
            "serialNumber": "000030169190",
            "vendor": "0001"
        },
        "log": {
            "components": {},
            "level": "warn",
            "storage": "volatile"
        },
        "network": {
            "authorized_keys": "",
            "interfaces": {
                "eth0": {
                    "ipAddressConfig": 0,
                    "ipv4": {
                        "address": "10.100.100.31",   
                        "dns": "255.255.255.0",       
                        "gateway": "10.100.100.1",    
                        "mask": 24
                    },
                    "mac": "48:B0:2D:68:66:72",       
                    "networkSpeed": 0,
                    "useDHCP": false
                },
                "eth1": {
                    "ipAddressConfig": 0,
                    "mac": "00:02:01:43:E6:D6",       
                    "networkSpeed": 0,
                    "useDHCP": true
                }
            }
        },
        "status": "OPERATE",
        "swVersion": {
            "euphrates": "1.18.3",
            "firmware": "0.16.23-823",
            "kernel": "4.9.140-l4t-r32.4+g8c7b68130d9a",
            "l4t": "r32.4.3",
            "schema": "v1.4.1",
            "tcu": "1.0.4"
        }
    },
    "ports": {
        "port0": {
            "acquisition": {
                "delay": 0,
                "framerate": 10.0,
                "version": {
                    "major": 0,
                    "minor": 0,
                    "patch": 0
                }
            },
            "data": {
                "algoDebugConfig": {
                    "version": {
                        "major": 0,
                        "minor": 13,
                        "patch": 12
                    }
                },
                "availablePCICOutput": [
                    "RGB_INFO"
                ],
                "pcicTCPPort": 50010
            },
            "info": {
                "calibration": {
                    "version": {
                        "major": 0,
                        "minor": 1,
                        "patch": 1
                    }
                },
                "device": "2301",
                "deviceTreeBinaryBlobOverlay": "001-ov9782.dtbo",
                "features": {
                    "fov": {
                        "horizontal": 127,
                        "vertical": 80
                    },
                    "resolution": {
                        "height": 800,
                        "width": 1280
                    },
                    "type": "2D"
                },
                "icc": {
                    "version": {
                        "major": 0,
                        "minor": 0,
                        "patch": 0
                    }
                },
                "name": "",
                "partNumber": "O3R225",
                "productionState": "AC",
                "sensor": "OV9782",
                "sensorID": "OV9782_127x80_noIllu_Csample",
                "serialNumber": "000031382187",       
                "vendor": "0001",
                "version": {
                    "major": 11,
                    "minor": 1,
                    "patch": 0
                }
            },
            "mode": "standard_autoexposure2D",        
            "processing": {
                "extrinsicHeadToUser": {
                    "rotX": 0.45,
                    "rotY": 0.0,
                    "rotZ": 0.0,
                    "transX": -0.15,
                    "transY": 1.74,
                    "transZ": -0.07
                },
                "version": {
                    "major": 0,
                    "minor": 13,
                    "patch": 12
                }
            },
            "state": "RUN"
        },
        "port1": {
            "acquisition": {
                "delay": 0,
                "framerate": 10.0,
                "version": {
                    "major": 0,
                    "minor": 0,
                    "patch": 0
                }
            },
            "data": {
                "algoDebugConfig": {
                    "version": {
                        "major": 0,
                        "minor": 13,
                        "patch": 12
                    }
                },
                "availablePCICOutput": [
                    "RGB_INFO"
                ],
                "pcicTCPPort": 50011
            },
            "info": {
                "calibration": {
                    "version": {
                        "major": 0,
                        "minor": 1,
                        "patch": 1
                    }
                },
                "device": "2301",
                "deviceTreeBinaryBlobOverlay": "001-ov9782.dtbo",
                "features": {
                    "fov": {
                        "horizontal": 127,
                        "vertical": 80
                    },
                    "resolution": {
                        "height": 800,
                        "width": 1280
                    },
                    "type": "2D"
                },
                "icc": {
                    "version": {
                        "major": 0,
                        "minor": 0,
                        "patch": 0
                    }
                },
                "name": "",
                "partNumber": "O3R225",
                "productionState": "AC",
                "sensor": "OV9782",
                "sensorID": "OV9782_127x80_noIllu_Csample",
                "serialNumber": "000030329739",       
                "vendor": "0001",
                "version": {
                    "major": 11,
                    "minor": 1,
                    "patch": 0
                }
            },
            "mode": "standard_autoexposure2D",        
            "processing": {
                "extrinsicHeadToUser": {
                    "rotX": 0.0,
                    "rotY": 0.0,
                    "rotZ": 0.0,
                    "transX": 0.0,
                    "transY": 0.635,
                    "transZ": 0.0
                },
                "version": {
                    "major": 0,
                    "minor": 13,
                    "patch": 12
                }
            },
            "state": "RUN"
        },
        "port2": {
            "acquisition": {
                "channelSelection": "manual",
                "channelValue": 5,
                "delay": 0,
                "exposureLong": 5000,
                "exposureShort": 400,
                "framerate": 20,
                "offset": 0.0,
                "swTriggerGroup": 0,
                "version": {
                    "major": 0,
                    "minor": 0,
                    "patch": 0
                }
            },
            "data": {
                "algoDebugConfig": {
                    "version": {
                        "major": 0,
                        "minor": 13,
                        "patch": 12
                    }
                },
                "availablePCICOutput": [
                    "TOF_INFO",
                    "RADIAL_DISTANCE_NOISE",
                    "RADIAL_DISTANCE_COMPRESSED",     
                    "REFLECTIVITY",
                    "AMPLITUDE_COMPRESSED",
                    "CONFIDENCE"
                ],
                "pcicTCPPort": 50012
            },
            "info": {
                "calibration": {
                    "version": {
                        "major": 0,
                        "minor": 6,
                        "patch": 1
                    }
                },
                "device": "3101",
                "deviceTreeBinaryBlobOverlay": "001-irs2381c.dtbo",
                "features": {
                    "fov": {
                        "horizontal": 105,
                        "vertical": 78
                    },
                    "resolution": {
                        "height": 172,
                        "width": 224
                    },
                    "type": "3D"
                },
                "icc": {
                    "version": {
                        "major": 1,
                        "minor": 0,
                        "patch": 6
                    }
                },
                "name": "",
                "partNumber": "O3R225",
                "productionState": "AC",
                "sensor": "IRS2381C",
                "sensorID": "IRS2381C_105x78_4x2W_110x90_C7",
                "serialNumber": "000031382187",       
                "vendor": "0001",
                "version": {
                    "major": 11,
                    "minor": 1,
                    "patch": 0
                }
            },
            "mode": "standard_range4m",
            "processing": {
                "diParam": {
                    "anfFilterSizeDiv2": 1,
                    "dynamicSymmetryThreshold": 10,   
                    "enableTemporalFilter": true,     
                    "maxDistNoise": 0.05,
                    "medianSizeDiv2": 2,
                    "minAmplitude": 10,
                    "minReflectivity": 0.1,
                    "mixedPixelFilterMode": 1,        
                    "mixedPixelThresholdRad": 0.5     
                },
                "extrinsicHeadToUser": {
                    "rotX": 0.0,
                    "rotY": 0.0,
                    "rotZ": 3.14,
                    "transX": 0.0,
                    "transY": 0.635,
                    "transZ": 0.015
                },
                "version": {
                    "major": 0,
                    "minor": 13,
                    "patch": 12
                }
            },
            "state": "RUN"
        },
        "port3": {
            "acquisition": {
                "channelSelection": "manual",
                "channelValue": 2,
                "delay": 0,
                "exposureLong": 5000,
                "exposureShort": 400,
                "framerate": 20,
                "offset": 0.0,
                "swTriggerGroup": 0,
                "version": {
                    "major": 0,
                    "minor": 0,
                    "patch": 0
                }
            },
            "data": {
                "algoDebugConfig": {
                    "version": {
                        "major": 0,
                        "minor": 13,
                        "patch": 12
                    }
                },
                "availablePCICOutput": [
                    "TOF_INFO",
                    "RADIAL_DISTANCE_NOISE",
                    "RADIAL_DISTANCE_COMPRESSED",     
                    "REFLECTIVITY",
                    "AMPLITUDE_COMPRESSED",
                    "CONFIDENCE"
                ],
                "pcicTCPPort": 50013
            },
            "info": {
                "calibration": {
                    "version": {
                        "major": 0,
                        "minor": 6,
                        "patch": 1
                    }
                },
                "device": "3101",
                "deviceTreeBinaryBlobOverlay": "001-irs2381c.dtbo",
                "features": {
                    "fov": {
                        "horizontal": 105,
                        "vertical": 78
                    },
                    "resolution": {
                        "height": 172,
                        "width": 224
                    },
                    "type": "3D"
                },
                "icc": {
                    "version": {
                        "major": 1,
                        "minor": 0,
                        "patch": 6
                    }
                },
                "name": "",
                "partNumber": "O3R225",
                "productionState": "AC",
                "sensor": "IRS2381C",
                "sensorID": "IRS2381C_105x78_4x2W_110x90_C7",
                "serialNumber": "000030329739",       
                "vendor": "0001",
                "version": {
                    "major": 11,
                    "minor": 1,
                    "patch": 0
                }
            },
            "mode": "standard_range4m",
            "processing": {
                "diParam": {
                    "anfFilterSizeDiv2": 1,
                    "dynamicSymmetryThreshold": 10,   
                    "enableTemporalFilter": true,     
                    "maxDistNoise": 0.05,
                    "medianSizeDiv2": 2,
                    "minAmplitude": 10,
                    "minReflectivity": 0.1,
                    "mixedPixelFilterMode": 1,        
                    "mixedPixelThresholdRad": 0.5     
                },
                "extrinsicHeadToUser": {
                    "rotX": 0.69,
                    "rotY": 0.0,
                    "rotZ": 3.14,
                    "transX": 0.0,
                    "transY": 1.775,
                    "transZ": -0.02
                },
                "version": {
                    "major": 0,
                    "minor": 13,
                    "patch": 12
                }
            },
            "state": "RUN"
        }
    }
}
`

**Comment by dekhanra** [comment]
Hey @Astronomia-maker, we were currently testing this scenarios but it would be great to have a Algo-Debug data from your test scenarios. Could you please send us a mail to support.robotics@ifm.com with data and we can follow up on this issue.

**Reply by Astronomia-maker** [reply]
Hi @dekhanra, is this correct one for algo-debug [Link](https://api.ifm3d.com/stable/examples/logging/logging.html)? Or you have any example script for python?

**Reply by dekhanra** [reply]
To record Algo Debug data on ifmVisionAssistant, follow the instructions below.

1. Open ifmVisionAssistant and connect to the sensor.
2. Click on `Monitor` window and navigate to Recording tab under the visualization window.
3. Select the duration of recording (Record for 5 seconds or 1 min)
4. Click on the `BUG` icon to record Algo-Debug  data
5. Finally click on `Record` icon to start the recording process.

![image](https://github.com/ifm/ifm3d/assets/123174684/7fa28779-23bb-468d-8d38-b5ba64587d04)
